### ISSUE TITLE
Docker Desktop Windows 2.2.0.0 breaks "nested" volume mounting

### ISSUE THREAD

**Issue by jk12559** [docker/for-win#5540]
Volumes mounted in a container which are inside a folder which is also a volume mount no longer have their content in the container.  See the reproduction steps below for an example.

###### Test Project Structure

* inner/
    * inner-file.txt (empty dummy file)
* outer/
    * outer-file.txt (empty dummy file)
* docker-compose.yml (see content)

###### docker-compose.yml
``` yml
version: "3.7"
services:
  centos:
    container_name: nested_mount_test
    image: centos
    command: sleep 10m
    volumes:
      - ./outer:/test
      - ./inner:/test/inner
```

1.	In the folder with docker-compose.yml file, run docker-compose up –d
2.	Run docker exec -it nested_mount_test /bin/bash
3.	Run cd /test and verify that there is a file called outer-file.txt and a folder called inner inside /test
4.	Observe the content of /test/inner

### Expected behavior
When observing the content of /test/inner, it should container inner-file.txt

### Actual behavior
When observing the content of /test/inner, it is empty

### Information
In Docker Desktop 2.1.0.4, this works as expected, but after upgrading to Docker Desktop 2.2.0.0, the volume mounted inside another volume has no content.

  - Windows Version: Windows 10 Version 1809
  - Docker Desktop Version: 2.2.0.0
  - Are you running inside a virtualized Windows e.g. on a cloud server or on a mac VM: No

**Comment by stephen-turner** [comment]
Thanks for the report. We have code that is supposed to handle this in 2.2.0.0, but maybe it's not fully working. Could you please supply a diagnostics file so that we can look into your logfiles?

**Comment by issacg** [comment]
#metoo

**Comment by issacg** [comment]
Unsure if this is related, but I'm getting a similar problem when trying to share a mount in a nested docker (using the host daemon)

Example:
1. On a directory in the host, create a file HOST.txt
2. `docker run --rm -ti -v /var/run/docker.sock:/var/run/docker.sock:ro -v %CD%:/data1 alpine sh`
3. 
```
/ # ls -al /data1
total 4
drwxrwxrwx    1 root     root             0 Jan 22 16:22 .
drwxr-xr-x    1 root     root          4096 Jan 22 16:24 ..
-rwxr-xr-x    1 root     root             0 Jan 22 16:22 HOST.txt
/ #
```
4. `# apk add docker-cli`
5.
```
/ # docker run --rm -ti -v /data1:/data2 alpine ls -al /data2
total 4
drwxr-xr-x    2 root     root            40 Jan 22 16:26 .
drwxr-xr-x    1 root     root          4096 Jan 22 16:26 ..
/ #
```

**Comment by houldsg** [comment]
We use this in all our node apps along with multi-stage builds for development, to keep the "node_modules" folder in the container and the source code on the host.

Dockerfile
```Dockerfile
FROM node:10-alpine AS dev-stage
WORKDIR /usr/src/app
COPY package*.json ./
RUN npm install -g npm@latest && npm install --no-audit

FROM dev-stage as testing-stage
COPY . /usr/src/app
```

docker-compose.yml
```yml
services:
  web:
    build:
      context: .
      dockerfile: Dockerfile
      target: dev-stage
    volumes:
      - .:/usr/src/app
      - /usr/src/app/node_modules
```

**Comment by cjancsar** [comment]
@houldsg Is the workflow you shared no longer working in 2.2? Or why did you share?

**Comment by houldsg** [comment]
@cjancsar Correct. The workflow is broken with 2.2.
It's the same "nested" volume mounting pattern except it's:
1. container (/)
2. host (/usr/src/app)
3. container again (/usr/src/app/node_modules)

**Comment by liayn** [comment]
From our POV this is a major blocker. Virtually all our projects use this feature.
Our usecase (simplified):

- ./html:/var/www/html
- 'path to some plugin':/var/www/html/plugins/pluginname/

**Comment by cjancsar** [comment]
@houldsg thanks, I think we are also experiencing that.

**Comment by zac-farrugia** [comment]
Also experiencing the same issue.

**Comment by dealer-solutions-gene** [comment]
What I found is the "child" volume would be created as an empty folder in the "parent" folder.

**Comment by dlford** [comment]
I'm seeing the same issue, same use case as @houldsg. I've had to downgrade to get some work done, but I did run diagnostics before doing so.

Diag ID: `7DB0481A-38FB-4150-A1C5-F752730ED428/20200123133355`

The volumes are created and populated (`node_modules`), I can mount the volumes in a new container and see the contents, but they don't work nested under the bind mounted project root folder.

Here is one configuration for reference:

`docker-compose.yml`

```yaml
version: '3.7'

  client_dev:
    volumes:
      - type: volume
        source: client-node_modules_dev
        target: /home/nextjs/app/node_modules
      - type: volume
        source: client-cache_dev
        target: /home/nextjs/app/.next
      - type: bind
        source: ./client
        target: /home/nextjs/app
    build:
      context: ./client
      dockerfile: Dockerfile.dev
    networks:
      net1:
        aliases:
          - client

networks:
  net1:
    name: project-name
volumes:
  client-node_modules_dev:
  client-cache_dev:
```

`Dockerfile`

```docker
FROM node:10-alpine

RUN mkdir -p /home/nextjs/app
WORKDIR /home/nextjs/app
COPY ./package*.json ./
RUN npm install

ENV NODE_ENV=development

EXPOSE 3000
CMD [ "npm", "run", "dev" ]
HEALTHCHECK --start-period=15s --interval=1m --timeout=5s CMD ["node", "healthcheck"]
```

**Comment by kybrdbnd** [comment]
> We use this in all our node apps along with multi-stage builds for development, to keep the "node_modules" folder in the container and the source code on the host.
> 
> Dockerfile
> 
> ```dockerfile
> FROM node:10-alpine AS dev-stage
> WORKDIR /usr/src/app
> COPY package*.json ./
> RUN npm install -g npm@latest && npm install --no-audit
> 
> FROM dev-stage as testing-stage
> COPY . /usr/src/app
> ```
> 
> docker-compose.yml
> 
> ```yaml
> services:
>   web:
>     build:
>       context: .
>       dockerfile: Dockerfile
>       target: dev-stage
>     volumes:
>       - .:/usr/src/app
>       - /usr/src/app/node_modules
> ```

messing my head from last 2 hours, and realized that its an issue

**Comment by simonferquel** [comment]
We have found a simpler repro case and are investigating the issue.
Repro is:
```
C:\Users\simon\repro> docker run --rm -v ~/:/test -v test-volume:/test/inner -it alpine touch /test/inner/data
C:\Users\simon\repro> docker run --rm -v ~/:/test -v test-volume:/test/inner -it alpine ls /test/inner/    
data
C:\Users\simon\repro> docker run --rm -v ~/:/test -v test-volume:/test/inner -it alpine sh                 
/ # ls /test/inner/
/ # exit
```
(see that the last one does not find the data file).

**Comment by pixelmixer** [comment]
Same issue here, seems pretty prevalent in 2.2.0.0. Started setting up a new local development workflow for a client and ran into this. Rolled back to an earlier version and it works, but isn't ideal getting a new workflow built on top of older version. Hopefully this one has some high priority. Thanks team!

**Comment by Herrick19** [comment]
We also have the same issue on all computers we updated… We'll roll back to the previous version while it's getting fixed

**Comment by Cyral** [comment]
Also experiencing this issue when attempting to keep the container's node_modules separate.

i.e.

```
volumes:
      - ./src/website:/home/app/website
      - /home/app/website/node_modules/
```

**Comment by krstffr** [comment]
Yep me too. 

Stopped updating docker for like six months due to issues like this always popping up after updates, guess I'll have to go back to that strategy again, very frustrating.

**Comment by cjancsar** [comment]
@Cyral 
> there is an unrelated issue with file watching

We were experiencing something similar to that as well and linked it with this issue. Our development watch-mode typescript compilation time skyrocketed in 2.2

**Comment by Cyral** [comment]
@cjancsar I can't figure out the watching issue and I haven't been able to find a github issue for it (or make a minimal example myself). _Some_ file watching is working but not all of it - I spent hours and am at a loss. May be a WSL issue and not docker related though.

@krstffr Same. And downgrading is a PITA - still spending this morning trying to find an older bug free version of Docker so we can get up and running again. These kinds of issues seem to be very common, both in our dev workflow with compose and our CI.

**Comment by stephen-turner** [comment]
Update: we think we've tracked this down, but it seems to be a Linux kernel bug(!), so we're being cautious about making the right change.

**Comment by EricHripko** [comment]
You can get an older version by uninstalling `2.2.0.0` and running `choco install docker-desktop --version 2.1.0.5`

**Comment by jskorlol** [comment]
Let's write a little more about this. 

#5592 
mount without problems for nginx
However, php-fpm will not mount.

Thank you.

**Comment by liayn** [comment]
And this shall be the direct download link for version 2.1.0.5
https://download.docker.com/win/stable/40693/Docker%20Desktop%20Installer.exe

**Comment by stephen-turner** [comment]
We have a new private build that (we believe) fixes this bug. https://download-stage.docker.com/win/stable/42579/Docker%20Desktop%20Installer.exe. Please try it and let us know how you get on.

**Comment by dealer-solutions-gene** [comment]
@stephen-turner I'm seeing the mounting working from the container point of view, the nested volume does reference the correct host destination. I do however see an artifact of an empty folder created in on the host machine representing the outer volume, but this may have been a preexisting behavior as I have seen rogue folders that aren't in my version control in my project which would explain them in a past version of the docker.

```
- ./proejct:/var/www/
- ./assets:/var/www/assets
```

With the above, an empty `./project/assets` folder is created on the host machine, any kind of interaction with this folder upsets the volume (eg deleting this empty folder or copying anything into it causes the files that were originally loaded via the `./assets` volume to start 404'ing under Apache)

**Comment by Cyral** [comment]
@stephen-turner Thanks for the fix, as far as I can tell it is now working for my configuration.

(For reference:)
```
volumes:
      - ./src/website:/home/app/website
      - /home/app/website/node_modules/ # This was previously not separated from the host volume
```

**Comment by LaurentLesle** [comment]
thanks for the fix. working for me as well

**Comment by dlford** [comment]
Working for me as well, thank you Docker team for all of your hard work.

**Comment by issacg** [comment]
@stephen-turner any hints on when this might be packaged inside a normal release?  Ballpark is fine (days, weeks, more)

**Comment by andriusjakas** [comment]
> We have a new private build that (we believe) fixes this bug. https://download-stage.docker.com/win/stable/42579/Docker%20Desktop%20Installer.exe. Please try it and let us know how you get on.

Has same problem, seems fixes

**Comment by stephen-turner** [comment]
Thanks, everyone.

@issacg Hopefully about a week, but it depends on the progress on #5530, which we don't yet have an in-house repro for.

**Comment by krstffr** [comment]
FYI the fix does not work for me, still not syncing.

**Comment by stephen-turner** [comment]
@krstffr Are you seeing the symptoms of this bug (where the inner directory was always empty) or the symptoms of #5530 (where a later change within the directory didn't reach the container)?

**Comment by krstffr** [comment]
@stephen-turner #5530

**Comment by stephen-turner** [comment]
Thanks, @krstffr. In that case I think this one is fixed. (I realise there are a lot of overlapping bugs here, with similar symptoms).

**Comment by stephen-turner** [comment]
We have just released 2.2.0.3, which fixes this issue, so closing this ticket. In case any Edge users are watching this ticket, it will be on the Edge channel tomorrow too.
